### PR TITLE
Draft design document for store

### PR DESCRIPTION
**Pull request by apostasie**
As requested by @fahedouch in #3364, this adds kind of a "design" document for the `store` implementation (still to be merged in 3364).

I am not sure this is the right tone, and I feel a bit bizarre explaining things that may (or may not) be obvious, so, please let me know your thoughts bluntly on whether any of this is helpful or not.

Furthermore, this PR creates a new `docs/dev` folder that is developer oriented (instead of user oriented) and moves the testing doc in there as well.

### CODE REVIEW

File: docs/dev/testing.md
Comment (AkihiroSuda) [inline docs/dev/testing.md:1]: Maybe this should be a separate commit

File: docs/dev/store.md [hunk 1]
```diff
@@ -0,0 +1,163 @@
+# Store
```
Comment (AkihiroSuda) [inline docs/dev/store.md:1]: "Store" needs explanation

Comment (apostasie) [reply docs/dev/store.md]: Amended that to more explicitly refer to `pkg/store`.

File: docs/dev/testing.md
Comment (apostasie) [reply docs/dev/testing.md:1]: Done.

**Review by AkihiroSuda — APPROVED** [review]
Thanks
